### ISSUE TITLE
Correct usage of 'formatters' and 'customFormatters' fields?

### ISSUE THREAD

**Issue by daldridge-cs** [angular-google-chart/angular-google-chart#186]
I have a chart type that I'm using in multiple views, and thus have wrapped it in a directive.  I'd like to use custom formatting on the annotation (abbreviated currency like '$1.2M') and tooltip (non-abbreviated currency).

I don't yet grok the usage of the 'formatters' and 'customFormatters' fields of chart object consumed by the google-chart directive (no documentation on it in angular-google-chart, nor does it appear to match anything on the Google Charts API documentation), and have instead resorted to waiting on the googleChartApiPromise to resolve, and use the charts API directly to create a DataView that specifies custom formatters.

Can anybody provide guidance on whether my goals can be accomplished through the use of the 'formatters' and/or 'customFormatters' fields (or something else I'm missing?), avoiding direct use of the visualization API (which somewhat negates the benefit of using this angular-google-charts module...) ?

(Note that my directive doesn't yet handle changing data, but that's a separate issue not relevant here.)

Thanks!
 
```javascript
'use strict';

angular.module('global').directive('csValueRiskChart', ['lodash', '$filter', 'googleChartApiPromise',
    function(lodash, $filter, googleChartApiPromise) {
        function roundWithPrecision(n, precision) {
            var prec = Math.pow(10, precision);
            return Math.round(n * prec) / prec;
        }

        function abbreviate(n) {
            var base = Math.floor(Math.log(Math.abs(n)) / Math.log(1000));
            var suffix = 'KMBT' [base - 1];
            return suffix ? roundWithPrecision(n / Math.pow(1000, base), 2) + suffix : '' + n;
        }

        function formatCurrency(column, dataTable, row) {
            var currency = $filter('currency');
            var dataTableValue = dataTable.getValue(row, column);
            return currency(dataTableValue);
        }

        function formatCurrencyShort(column, dataTable, row) {
            var dataTableValue = dataTable.getValue(row, column);
            return '$' + abbreviate(dataTableValue);
        }

        function preLink(scope, element, attrs) {
            var totalValue = lodash.sum(scope.claims, 'ClaimAmount');
            var riskValue = lodash.sum(scope.claims, 'ValueAtRisk');

            googleChartApiPromise.then(function() {
                var dataTable = window.google.visualization.arrayToDataTable([
                    ['Series', {
                        label: 'Amount',
                        type: 'number'
                    }, {
                        role: 'style'
                    }],
                    ['Total Value', totalValue, '#AFCE35'],
                    ['Value at Risk', riskValue, '#ee2223']
                ]);

                var dataView = new window.google.visualization.DataView(dataTable);
                dataView.setColumns([0, 1, 2, {
                    calc: formatCurrencyShort.bind(undefined, 1),
                    sourceColumn: 1,
                    type: 'string',
                    role: 'annotation'
                }, {
                    calc: formatCurrency.bind(undefined, 1),
                    sourceColumn: 1,
                    type: 'string',
                    role: 'tooltip'
                }]);
                
                scope.chartObject = {
                    type: 'ColumnChart',
                    options: {
                        vAxis: {
                            format: 'short'
                        }
                    },
                    formatters: {},
                    data: dataView
                };
            });
        }

        return {
            restrict: 'E',
            replace: true,
            scope: {
                claims: '='
            },
            templateUrl: '/assets/modules/global/views/valuerisk.client.view.html',
            compile: function(element, attrs) {
                return {
                    pre: preLink
                };
            }
        };
    }
]);
```

**Comment by Balrog30** [comment]
Yes, Formatters are a very poorly documented feature of Angular-Google-Chart.  I'll see what I can do to help you out here.  When I first started using the directive I actually did some refactoring on the code just to understand how they work.

Add something like this to your code (this formatter taken from my own project applies a date format like "Jun 8, 2015 (2015W24)", with "W24" being the 24th week of the year.

```JavaScript
$scope.myChart = {
  type: "LineChart",
  data: [] //your data here,
  options: {},
  formatters: "date": [{
    columnNum: 0, // column index to apply format to
    pattern: "MMM d, yyyy '('yyyy'W'w')'"
  }]
};
```
See the google documentation [here](https://developers.google.com/chart/interactive/docs/reference#formatters).

Please note that Formatters cannot be applied to a `DataView` as they are **Read Only**.  However, you can apply a View *and* a Table to a chart.

```JavaScript
$scope.myChart = {
  type: "LineChart",
  data: dataTable,
  view: {columns: [0,1,2,3]}
};
```

I do something like that in my blog post [here](http://nicholasbering.ca/angularjs/2014/08/16/agc-hide-series/), where I use this technique to show and hide series on a line graph my replacing them with calculated columns that return nothing.

CustomFormatters is a little more complicated, but basically you implement a JavaScript Constructor to build a formatter conforming to Google's interface (it takes params in the constructor, then has a `.format(dataTable, columnIndex)` function).  You add that to customFormatters (I think it works as a dictionary with the key being your formatter's name as it will be used by the formatters settings dictionary.  Then, you add the settings to the formatters list like I did above for date formatting and angular-google-chart takes care of the rest.

If you have multiple charts using the same data table instance, but do not want them all formatted, you'll need to clone it or something so that you aren't actually using the same instance anymore.  If you're passing your data as an array, it shouldn't be a problem.

I hope this helps, I didn't have a lot of time to edit this into a nice pretty documentation file.
